### DISCUSSION TITLE
[ZK Stack] - Difficulties deploying L2 chain to Sepolia

### DISCUSSION THREAD

**Discussion by bend-n** [zkSync-Community-Hub/zksync-developers#838] [ZK Stack]
### Module Affected

- [ ] System contracts (era-system-contracts)
- [ ] Smart contracts (era-contracts)
- [X] zkEVM (era-zk_evm)
- [X] Server (zksync-era)
- [ ] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [ ] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [ ] Other (please specify below)

### Rust Version

1.81.0-nightly

### Operating System

- [ ] Ubuntu
- [ ] MacOS
- [ ] Windows
- [X] Other (Please specify below)

### Issue Description

When I run `zkstack ecosystem init --l1-rpc-url 'https://rpc.sepolia.org/'`, and attempt to register the chain, it says:
```
│      Deploying Hyperchain
│      Governance deployed at: 0x1b8e729E36124923185dc3A5194282F60971d5Fc
│      ChainAdmin deployed at: 0x4C427a0651BD716c252D68FE5dDE092fe8dC5384
│      Token already registered on Bridgehub
|
│    Stderr:
│      Error: script failed: revert: Ownable: caller is not the owner
│      
■  Command failed to run: forge script deploy-scripts/RegisterHyperchain.s.sol --legacy --ffi --rpc-url=https://rpc.sepolia.org/ --broadcast --private-key=REDACTED
│  Error: script failed: revert: Ownable: caller is not the owner
```

### Expected Behavior

Expected to deploy the hyperchain.

### Current Behavior

Long error.

### Additional Details

stack trace
```
│      Traces:
│      [4474686] → new RegisterHyperchainScript@0x5b73C5498c1E3b4dbA84de0F1833c4a029d90519
│      └─ ← [Return] 22237 bytes of code
│      
│      [4866482] RegisterHyperchainScript::run()
│      ├─ [0] console::log("Deploying Hyperchain") [staticcall]
│      │ └─ ← [Stop]
│      ├─ [0] VM::projectRoot() [staticcall]
│      │ └─ ← [Return] "/home/os/zksync-era/contracts/l1-contracts"
│      ├─ [0] VM::readFile("/home/os/zksync-era/contracts/l1-contracts/script-config/register-hyperchain.toml")
│      [staticcall]
│      │ └─ ← [Return] <file>
│      ├─ [0] VM::parseTomlAddress("<stringified TOML>", "$.owner_address") [staticcall]
│      │ └─ ← [Return] 0xaddr
│      ├─ [0] VM::parseTomlAddress("<stringified TOML>", "$.deployed_addresses.bridgehub.bridgehub_proxy_addr")
│      [staticcall]
│      │ └─ ← [Return] 0x35A54c8C757806eB6820629bc82d90E056394C92
│      ├─ [0] VM::parseTomlAddress("<stringified TOML>",
│      "$.deployed_addresses.state_transition.state_transition_proxy_addr") [staticcall]
│      │ └─ ← [Return] 0x4e39E90746A9ee410A8Ce173C7B96D3AfEd444a5
│      ├─ [0] VM::parseTomlAddress("<stringified TOML>", "$.deployed_addresses.validator_timelock_addr") [staticcall]
│      │ └─ ← [Return] 0xD3876643180A79d0A56d0900C060528395f34453
│      ├─ [0] VM::parseTomlBytes("<stringified TOML>", "$.contracts_config.diamond_cut_data") [staticcall]
│      │ └─ ← [Return]
...
│      ├─ [0] VM::parseTomlUint("<stringified TOML>", "$.chain.chain_chain_id") [staticcall]
│      │ └─ ← [Return] 270
│      ├─ [0] VM::parseTomlUint("<stringified TOML>", "$.chain.bridgehub_create_new_chain_salt") [staticcall]
│      │ └─ ← [Return] 5846041104530926066 [5.846e18]
│      ├─ [0] VM::parseTomlAddress("<stringified TOML>", "$.chain.base_token_addr") [staticcall]
│      │ └─ ← [Return] ECRecover: [0x0000000000000000000000000000000000000001]
│      ├─ [0] VM::parseTomlBool("<stringified TOML>", "$.chain.validium_mode") [staticcall]
│      │ └─ ← [Return] false
│      ├─ [0] VM::parseTomlAddress("<stringified TOML>", "$.chain.validator_sender_operator_commit_eth") [staticcall]
│      │ └─ ← [Return] 0xaddr
│      ├─ [0] VM::parseTomlAddress("<stringified TOML>", "$.chain.validator_sender_operator_blobs_eth") [staticcall]
│      │ └─ ← [Return] addr
│      ├─ [0] VM::parseTomlUint("<stringified TOML>", "$.chain.base_token_gas_price_multiplier_nominator") [staticcall]
│      │ └─ ← [Return] 1
│      ├─ [0] VM::parseTomlUint("<stringified TOML>", "$.chain.base_token_gas_price_multiplier_denominator") [staticcall]
│      │ └─ ← [Return] 1
│      ├─ [0] VM::parseTomlUint("<stringified TOML>", "$.chain.governance_min_delay") [staticcall]
│      │ └─ ← [Return] 0
│      ├─ [0] VM::parseTomlAddress("<stringified TOML>", "$.chain.governance_security_council_address") [staticcall]
│      │ └─ ← [Return] 0x0000000000000000000000000000000000000000
│      ├─ [0] VM::broadcast()
│      │ └─ ← [Return]
│      ├─ [1062045] → new Governance@0x1b8e729E36124923185dc3A5194282F60971d5Fc
│      │ ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner:
│      addr)
│      │ ├─ emit OwnershipTransferred(previousOwner: addr, newOwner:
│      addr)
│      │ ├─ emit ChangeSecurityCouncil(_securityCouncilBefore: 0x0000000000000000000000000000000000000000,
│      _securityCouncilAfter: 0x0000000000000000000000000000000000000000)
│      │ ├─ emit ChangeMinDelay(_delayBefore: 0, _delayAfter: 0)
│      │ └─ ← [Return] 5127 bytes of code
│      ├─ [0] console::log("Governance deployed at:", Governance: [0x1b8e729E36124923185dc3A5194282F60971d5Fc])
│      [staticcall]
│      │ └─ ← [Stop]
│      ├─ [0] VM::broadcast()
│      │ └─ ← [Return]
│      ├─ [679704] → new ChainAdmin@0x4C427a0651BD716c252D68FE5dDE092fe8dC5384
│      │ ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner:
│      addr)
│      │ ├─ emit OwnershipTransferred(previousOwner: addr, newOwner:
│      0xaddr)
│      │ ├─ emit NewTokenMultiplierSetter(_oldTokenMultiplierSetter: 0x0000000000000000000000000000000000000000,
│      _newTokenMultiplierSetter: 0x0000000000000000000000000000000000000000)
│      │ └─ ← [Return] 3235 bytes of code
│      ├─ [0] console::log("ChainAdmin deployed at:", ChainAdmin: [0x4C427a0651BD716c252D68FE5dDE092fe8dC5384])
│      [staticcall]
│      │ └─ ← [Stop]
│      ├─ [9716] 0x35A54c8C757806eB6820629bc82d90E056394C92::tokenIsRegistered(ECRecover:
│      [0x0000000000000000000000000000000000000001]) [staticcall]
│      │ ├─ [2621] 0x0029e562c0b54C0b88cB22adF4346DbfEC87400c::tokenIsRegistered(ECRecover:
│      [0x0000000000000000000000000000000000000001]) [delegatecall]
│      │ │ └─ ← [Return] true
│      │ └─ ← [Return] true
│      ├─ [0] console::log("Token already registered on Bridgehub") [staticcall]
│      │ └─ ← [Stop]
│      ├─ [0] VM::recordLogs()
│      │ └─ ← [Return]
│      ├─ [3014] 0x35A54c8C757806eB6820629bc82d90E056394C92::admin() [staticcall]
│      │ ├─ [2422] 0x0029e562c0b54C0b88cB22adF4346DbfEC87400c::admin() [delegatecall]
│      │ │ └─ ← [Return] 0x1112c822C6E05399cb526509564449FDb6D4De14
│      │ └─ ← [Return] 0x1112c822C6E05399cb526509564449FDb6D4De14
│      ├─ [0] VM::broadcast()
│      │ └─ ← [Return]
│      ├─ [2717] 0x1112c822C6E05399cb526509564449FDb6D4De14::multicall([Call({ target:
│      0x35A54c8C757806eB6820629bc82d90E056394C92, value: 0, data:
│      0x3f58f....
│      })], true)
│      │ └─ ← [Revert] revert: Ownable: caller is not the owner
│      └─ ← [Revert] revert: Ownable: caller is not the owner
│
│      Deploying Hyperchain
│      Governance deployed at: 0x1b8e729E36124923185dc3A5194282F60971d5Fc
│      ChainAdmin deployed at: 0x4C427a0651BD716c252D68FE5dDE092fe8dC5384
│      Token already registered on Bridgehub
|
│    Stderr:
│      Error: script failed: revert: Ownable: caller is not the owner
│      
```

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by alhenaai** [comment]
It appears that the error you're encountering, "Ownable: caller is not the owner," typically occurs when a contract function that is restricted to the contract's owner is called by an address that is not recognized as the owner. This can happen if the private key used does not match the owner's address set in the contract's state or if there's a misconfiguration in the deployment scripts.

Here are a few steps you can take to troubleshoot and resolve this issue:

1. **Verify Owner Address**: Ensure that the private key used for the deployment corresponds to the owner address expected by the contract. You might need to check the deployment configuration or the initial setup of the contract to confirm the owner's address.

2. **Contract Deployment Review**: Review the deployment scripts, especially `RegisterHyperchain.s.sol`, to ensure that they are using the correct owner address during the deployment process.

3. **Configuration Files**: Check any configuration files (like TOML files mentioned in your stack trace) to ensure that the owner address and other parameters are correctly set.

4. **Manual Verification**: If possible, manually verify the ownership by interacting with the contract using a script or a tool like Hardhat or Truffle to call the `owner()` function of the deployed contract to see what address is returned.

If these steps do not resolve the issue, you may need to redeploy the contracts with the correct configurations or seek further assistance from the development team or community. For more detailed discussions and support, you can also visit the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by JackHamer09** [comment]
Hey, thanks for submitting the issue. I'll investigate and get back to you ASAP!

**Comment by kaymomin** [comment]
@bend-n did you try deploying it locally, did it work in that environment? I'm checking with the team further about this issue.

**Reply by bend-n** [reply]
You mean just the `Localhost` mode? Actually no, with `--dev` mode, after creating the ecosystem for `Localhost`, it says
```
⚙  Command completed: forge build --zksync --zk-enable-eravm-extensions
│  
◇  Deploying ecosystem contracts...
│
■  Deserialization Error: expected value at line 1 column 1. Response: Authorization header is missing or invalid
```

**Comment by sanekmelnikov** [comment]
Right now, chain registration within ZKsync ecosystem is permissioned. Meaning to add a newly created chain on Sepolia to already existing BridgeHub ([0x35A54c8C757806eB6820629bc82d90E056394C92](https://sepolia.etherscan.io/address/0x35A54c8C757806eB6820629bc82d90E056394C92 )) you'll need a signature of ecosystem admin which is governed by Matter Labs, hence the error "caller is not the owner".
If you want to use it for testing, you can deploy your own ecosystem on Sepolia and register your new chain within your own ecosystem. That will deploy your version of BridgeHub and other ecosystem contracts, you'll have a total control of.
In order to do that, during `zkstack ecosystem init` step, you need to provide "Yes" as an answer to this question:
> `Do you want to deploy ecosystem contracts? (Not needed if you already have an existing one)`
> `Yes`

**Reply by bend-n** [reply]
But i did?

**Reply by bend-n** [reply]
I believe part of the issue was the sepolia rpc— after replacing that, it has an error saying: 
![x](https://github.com/user-attachments/assets/895a82e1-bd8e-4ace-9ba5-1c066de82c68)

**Reply by bxpana** [reply]
Try running `zkstack ecosystem init --resume -v` to see if you can pick up where you left off and try again. I just setup a fresh ecosystem and chain on Sepolia today and was able to complete everything successfully

**Reply by bend-n** [reply]
with that it says 
![image](https://github.com/user-attachments/assets/d24af36f-0748-4a60-b716-186424330163)
also, just trying the whole thing again seems to usually get the same result.

**Reply by bxpana** [reply]
And you get the same error of the address already validator?

**Reply by bend-n** [reply]
you mean when i try the whole thing again? yeah

**Reply by bxpana** [reply]
When you try again are you starting fresh with `zkstack ecosystem create`?

What OS are you using?

**Comment by popzxc** [comment]
I'm going to close this discussion as stale; @bend-n if you didn't resolve the issue, please start a new discussion.
